Adding flag to disable opening browser window

hooksie1 opened the pull request:
This PR disables opening the browser window automatically for SSO.

To disable the browser just pass `--disable-browser` and it only
prints the URL.

hooksie1 committed:
Adding flag to disable opening browser window

This PR disables opening the browser window automatically for SSO.

To disable the browser just pass `--disable-browser` and it only
prints the URL.

gravitational-jenkins commented:
Can one of the admins verify this patch?

webvictim committed:
Merge branch 'master' into issue-3321

webvictim commented:
ok to test

russjones commented:
@benarent What do you think about this functionality and UX? If you give a :+1: we can do a code review.

benarent commented on tool/tsh/tsh.go:
```diff
@@ -268,6 +272,7 @@ func Run(args []string, underTest bool) {
 		identityfile.FormatOpenSSH)).Default(string(identityfile.DefaultFormat)).StringVar((*string)(&cf.IdentityFormat))
 	login.Flag("request-roles", "Request one or more extra roles").StringVar(&cf.DesiredRoles)
 	login.Arg("cluster", clusterHelp).StringVar(&cf.SiteName)
+	login.Flag("disable-browser", noBrowserHelp).BoolVar(&cf.NoBrowser)
```
I think we should add this feature but change the CLI experience.

I looked around, and I would suggest that we add a `--open` flag, this can then be extended in the future to add  `--browserTarget=browserTarget`


The new Flag will be `--open=false`

### TSH Login without auto-open of Browser.  

```
tsh login --proxy=teleport.example.com --open=false 
```

webvictim replied:
How about make it `--browser=none` to not open a browser at all? That way we can extend this in future to make use of things like `--browser=chrome` etc.

benarent replied:
yeah, that's a very elegant solution. I've updated the question to suggest it.

webvictim committed:
Implement --browser flag to tsh login

(cherry picked from commit edf1ddf228881afc6cfeb1226c4f208a776831f7)

webvictim commented:
I implemented the fixes as described here on a branch: https://github.com/gravitational/teleport/commits/hooksie1/issue-3321

(also rebased this against `master` and fixed the conflicts)

hooksie1 commented:
Hey guys, sorry I haven't been responsive and thanks for updating that @webvictim! I had surgery last week so I've been kind of out of everything and wasn't keeping up with the discussion on here.

webvictim commented:
@hooksie1 No problem! I wasn't intending to be pushy or anything, I just happened to be looking at the codebase and saw the opportunity to commit a fix and handle the rebase. Feel free to cherry-pick my commit into your branch or whatever you like :)

hooksie1 commented:
@webvictim you weren't being pushy at all! It was a genuine thanks for doing the work!

hooksie1 commented:
Ok I think I did that right. Admittedly it was my first cherry-pick lol.

webvictim committed:
Merge branch 'master' into issue-3321

webvictim commented:
@gravitational-jenkins retest this please

webvictim commented:
@benarent It looks like we're happy with the UX now, so maybe we can move to code review?

benarent commented on lib/client/api.go:
```diff
@@ -253,8 +253,9 @@ type Config struct {
-	// NoBrowser will disable automatically opening the browser window for SSO login.
-	NoBrowser bool
+	// Browser can be used to pass the name of a browser to override the system default
+	// (not currently implemented), or set to 'none' to suppress browser opening entirely.
```
👍 Good to know that the ground work has been done but doesn't need extra docs yet.

webvictim commented:
@russjones @awly @fspmarshall Please take a look.

fspmarshall reviewed (approved):
LGTM

awly commented on lib/client/weblogin.go:
```diff
@@ -335,7 +335,7 @@ func Find(ctx context.Context, proxyAddr string, insecure bool, pool *x509.CertP
 
 // SSHAgentSSOLogin is used by tsh to fetch user credentials using OpenID Connect (OIDC) or SAML.
-func SSHAgentSSOLogin(login SSHLogin) (*auth.SSHLoginResponse, error) {
+func SSHAgentSSOLogin(login SSHLogin, tc *TeleportClient) (*auth.SSHLoginResponse, error) {
```
since this func doesn't actually make any teleport requests, pass in just the browser string instead of the entire `TeleportClient`

hooksie1 replied:
Ok done.

awly commented on lib/client/weblogin.go:
```diff
@@ -354,29 +354,34 @@ func SSHAgentSSOLogin(login SSHLogin) (*auth.SSHLoginResponse, error) {
 	case teleport.DarwinOS:
 		path, err := exec.LookPath(teleport.OpenBrowserDarwin)
-		if err == nil {
+		if err == nil && tc.Config.Browser != teleport.BrowserNone {
```
surround the entire `swich` statement with `if tc.Config.Browser != teleport.BrowserNone` instead, to avoid the unnecessary `$PATH` lookup

hooksie1 replied:
Ok done.

hooksie1 committed:
Don't pass in full client and wrap if statement

Instead of sending the whole tc.Client in to SSHAgentSSOLogin,
just passing the browser string.

Also removed the separate if checks for the browser string and
wrapped the whole switch statement checking for browser != none.

awly commented on lib/client/weblogin.go:
```diff
@@ -350,33 +350,40 @@ func SSHAgentSSOLogin(login SSHLogin) (*auth.SSHLoginResponse, error) {
-		path, err := exec.LookPath(teleport.OpenBrowserLinux)
-		if err == nil {
-			execCmd = exec.Command(path, clickableURL)
+	if browser != "none" {
```
nit: `if browser != teleport.BrowserNone {`

awly commented on lib/client/weblogin.go:
```diff
@@ -350,33 +350,40 @@ func SSHAgentSSOLogin(login SSHLogin) (*auth.SSHLoginResponse, error) {
-	fmt.Printf("clicking on the link:\n %v\n", clickableURL)
+	// Print the URL to the screen, in case the command that launches the browser did not run.
+	// If Browser is set to the special string teleport.BrowserNone, no browser will be opened.
+	if browser == "none" {
```
nit: `if browser == teleport.BrowserNone {`

hooksie1 replied:
Sorry maybe I'm missing something. BrowserNone changed to just Browser to incorporate the string. It was BrowserNone when it was a bool. I just noticed though, I didn't change the comment above the if statement to reflect that.

hooksie1 replied:
Nope I'm dumb. Nevermind, I didn't see the const value.

hooksie1 committed:
Cleaned up values

Used the const teleport.BrowserNone instead of just directly
checking for "none".
